### ISSUE TITLE
jQuery UI alternative if needed

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#1398]
We currently use parts of jQuery UI in our backend, for UI items such as:

* Modal windows
* Dropdown lists
* Tabs
* Controlgroups
* A few other UI bits, such as datepicker, is available to plugin authors.

However development of jQuery UI has all but stalled and I'm concerned the project will become unsuitable for our needs going forward (lack of touch enabled UI elements is a problem, for example).

So, in light of this, maybe we should start investigating alternative JavaScript libraries that give us the UI elements above amongst others? Especially if we are writing a new widgets interface anyway.

If you have any suggestions then please note them below. Off the top of my head here are a few options (note that I am ruling out React from this list).

* [Vue.js](https://vuejs.org) (MIT license)
* [UIkit](https://getuikit.com) (MIT license)
* [Bootstrap](https://getbootstrap.com) (just the JavaScript bits) (MIT license)

**Comment by Bloke** [comment]
I don't  have any strong preference. Fine with whichever is the smallest, most lightweight and fastest that provides us the tools we need. And the one that causes the least headaches to integrate.

> note that I am ruling out React from this list

Amen to that. I've tried it and fail to see what all the fuss is about. Pretty much every design/dev job I've seen advertised in the last year or so has stipulated proficiency in React as either an essential or desirable skill. It's slow, bloated and transformed every page into the equivalent of wading through a river of treacle. No thanks.

**Comment by bloatware** [comment]
From a brief look, it seems that all of these (incl. jQuery UI) use their own markup, which makes switching from one to another a pain. We have already a bunch of weirdly looking `ui-` selectors in our code. I would rather investigate vanilla js solutions first.

**Comment by Bloke** [comment]
True. JS itself has come on massively in recent years so maybe native code is viable now.

**Comment by philwareham** [comment]
Yeah, sure. I'm all for vanilla JS if and where we can. My original concern is that jQuery UI development is pretty much dead now so we need to factor that into our future plans. And it makes sense to do that during the widget refresh.

Luckily we only currently use a few bits from jQuery UI, so we don't need to write a whole library of UI bits ourselves - just the bits we use.

**Comment by bloatware** [comment]
We don't necessary need to write a library, we (me) should just avoid using third-party selectors as much as possible to make migrations easier. Vue.js looks interesting because of its `{{template}}` features we could use in message panes, file download previews etc.

**Comment by Bloke** [comment]
Agree. Perhaps when we refactor `textpattern.js` I think we should maybe take the time to replace `$(.selector)` with standard JavaScript selectors now that the mechanisms to do so are more readily available. Start to wean ourselves off jQuery reliance.

Not sure of the best one (`querySelector*` I guess) but in any case, [traversing the DOM](https://zellwk.com/blog/dom-traversals/) seems a lot easier now in vanilla JS.

**Comment by philwareham** [comment]
My brother's company use Vue.js after testing a lot of alternatives - including React - and they think it's great. So maybe worth testing further.

**Comment by Sacripant** [comment]
You made the choice a few years ago not to use a php framework for Textpattern. His weakness, his advantage, his particularity. 
I will suggest you to stay in the same strategy for core JS: ES6 + BabelJS compilation for old browser support. Don't be tempted by the framework JS phenomenon, your team is too small to be able to follow them. 
https://dev.to/winduptoy/a-javascript-free-frontend-2d3e
But your choice will probably depend the needs for future improvements of the admin interface. in particular custom fields and possibility to create custom write tab UI per section or content types (BWT in core).

**Comment by bloatware** [comment]
We could make a js-less admin interface (like it was originally), but we'd loose

- the possibility to undo/redo changes when saving an article
- large files upload
- local UI state storage
- many dynamic UI enhancements

Some of them can and should be coded in vanilla js/css, but we can not get rid of jQuery and some file upload script -- just too big to fall.

**Comment by Bloke** [comment]
I do understand people's motivations for keeping things with as little external dependency as possible. But removing JS components entirely would cripple the UX.

jQuery is (was?) convenient at the time for cross-browser support. But perhaps now that ES6 has matured and is way more capable, and our target browser set all run the latest version, if we audit our usage of jQuery and find that we're only using a fraction of its feature set, maybe designing it out of the equation completely and refactoring everything as native JS is possible.

The only dependency we may have is the choice of upload script if it's a jQuery plugin. In which case we may be able to either build a smaller jQuery package that only includes the features we need, or find an upload script that's written in vanilla JS and use that instead.

**Comment by philwareham** [comment]
Regarding a js-less interface, there are also serious accessibility limitations from removing JavaScript entirely. Setting correct WAI-ARIA states is just not possible without some JS here and there. In the comments section of the article @Sacripant links to the author admits he knows very little about UI accessibility.

I think what we are all agreed on is that we need to modernise, simplify and reduce JavaScript where feasible - and use vanilla JS in preference to a third-party library where possible (although we need to balance that out with the time and effort it takes to make our own UI widgets from scratch). The admin panel is always going to need JavaScript.

**Comment by philwareham** [comment]
In the short-term, we can slim down the jQuery UI library elements to only the bits we actually use - that will make transitioning to a new library or writing our own code a simpler task.

Please tick any items we **cannot omit** because they are in use currently:

[ ] Interactions: Draggable
[ ] Interactions: Resizable
[ ] Interactions: Selectable
[ ] Interactions: Sortable
[ ] Widgets: Accordion
[ ] Widgets: Autocomplete
[ ] Widgets: Button
[ ] Widgets: Checkboxradio
[ ] Widgets: Controlgroup
[ ] Widgets: Datepicker
[ ] Widgets: Dialog
[ ] Widgets: Menu
[ ] Widgets: Progressbar
[ ] Widgets: Selectmenu
[ ] Widgets: Slider
[ ] Widgets: Spinner
[ ] Widgets: Tabs
[ ] Widgets: Tooltip
[ ] Styles: Icons

**Comment by bloatware** [comment]
If we ever switch to `Vue.js`, we'll be able to plug in [tiptap](https://tiptap.scrumpy.io/) editor based on ProseMirror, but seemingly more easily configurable.

**Comment by philwareham** [comment]
Shall we do a branch where we can test this and generally improve JS throughout?

**Comment by bloatware** [comment]
I think yes, but probably later, when we get a more clear idea? We have already 3 branches in work.

**Comment by bloatware** [comment]
> My brother's company use Vue.js after testing a lot of alternatives - including React - and they think it's great. So maybe worth testing further.

After some testing I find it great too - and rather easy to integrate. The way data is bound to views is very cute: the view is automatically updated as soon as the associated data changes. This allows for a nice data/view separation, unlike our current implementation, and also code simplification.

The price to pay is a proprietary markup, which makes orphans all plugins using `ui-` callbacks. But if at some stage we have to do not bwc changes anyway, Vue.js is worth considering.

**Comment by philwareham** [comment]
I wonder if we could have some sort of cross-over/migration period where the older UI is still available - is that even possible? We could strip jQuery UI back to the elements ticked in the list above to reduce legacy code size.

Or... we decide that we are definitely doing next release as Textpattern 5 and write modern code no (along with Stef's new widgets code).

**Comment by bloatware** [comment]
UI is not the only question - our entire model would change. Say, currently we send (via async request) the whole markup to update the `ID` link on the articles list tab:
```
<a href="?event=article&step=edit&ID=41">41</a>
```
Plugins can alter this block via `pluggable_ui` callbacks.

With Vue.js the template would be
```
<a :href="'?event=article&step=edit&ID='+ID">{{ID}}</a>
```
and we would send the data separately, like `"ID": "41"`. This means that

1. current plugins would stop working
2. to fix them, authors need to know Vue.js markup
3. an eventual further switching from Vue.js to another framework is cumbersome

So it would definitely be txp 5 for me.

**Comment by philwareham** [comment]
Just putting this here for future reference too:

https://github.com/ElemeFE/element

UI component library built on Vue.js.

Also a sidenote, Due has built-in drag/drop page reordering stuff so might be worth investigating for custom fields plans.

**Comment by philwareham** [comment]
On the other hand - the JS components from UIkit would probably be the path of least resistance, since it doesn't involve a complete rethink of how our admin side works.

**Comment by Bloke** [comment]
JS has come on massively, as that article states. Definitely time to design it out of our systems when we overhaul textpattern.js.

**Comment by petecooper** [comment]
Follow up to https://github.com/textpattern/textpattern/issues/1398#issuecomment-1131860132 –

https://insidegovuk.blog.gov.uk/2022/08/15/the-impact-of-removing-jquery-on-our-web-performance/

**Comment by philwareham** [comment]
I've been using UIkit as part of a client job (it was already in place) and I like it, has replacements for all the useful jQuery UI things (and more besides, obviously) and markup wouldn't need to change that much to incorporate it (but yes, markup would need changes and new classnames, although the majority f our existing CSS could be adapted quite easily to fit it).

It's also mature, has good support and lots of accessibility built in. Will investigate a Textpattern integration and see how it pans out.

**Comment by Bloke** [comment]
If it's easier to sidestep to something well supported (and we can eventually remove jQuery itself from core?) as we move towards more native JS where that makes sense, then that's fine with me. Seems like attribute name hooks just need a `data-uk-`prefix and the classnames of the components are broadly compatible with fairly minimal tweaking.

If it works out, we'll deprecate jQuery UI (and jQuery when we've refactored all the `$()` stuff) for a few versions and eventually axe it.
